modules/assim.sequential/R/Analysis_sda.R

```diff
@@ -0,0 +1,415 @@
+                       Observed=list(R=NULL,Y=NULL),
+                       H,
+                       ...
+){
+  
+  if (is.null(FUN)) PEcAn.logger::logger.severe('Analysis function needs to be defined !')
+  FUN(settings, Forecast, Observed, H, ...)
+  
+}
+
+##' @title EnKF
+##' @name  EnKF
+##' @author Michael Dietze \email{dietze@@bu.edu}, Ann Raiho and Hamze Dokoohaki
+##' 
+##' @param settings  pecan standard settings list.  
+##' @param Forecast A list containing the forecasts variables including Pf (cov of forecast state variables), mu.f (vector of estimated mean of forecast state variables),
+##' Q (process variance) and X (a dataframe of forcats state variables for different ensemble)
+##' @param Observed A list containing the observed variables including R (cov of observed state variables) and Y (vector of estimated mean of observed state variables)
+##' @param ... Extra argument sent to the analysis function.
+##’ @details
+##’  
+##' 
+##' @description Given the Forecast and Observed this function performs the Ensemble Kalamn Filter. 
+##' 
+##' @return It returns a list with estimated mean and cov matrix of forecast state variables as well as mean and cov estimated as a result of assimilation/analysis .
+##' @export
+EnKF<-function(setting, Forecast, Observed, H, ...){
+  
+  #------------------------------Setup
+  #-- reading the dots and exposing them to the inside of the function
+  dots<-list(...)
+  if (length(dots)>0) lapply(names(dots),function(name){assign(name,dots[[name]])})
+  for(i in seq_along(dots)) assign(names(dots)[i],dots[[names(dots)[i]]])
+
+    #Forecast inputs 
+  Q <- Forecast$Q # process error
+  Pf <- Forecast$Pf # Forecast precision
+  mu.f <- Forecast$mu.f #mean Forecast
+  X <- Forecast$X # states 
+  #Observed inputs
+  R <- Observed$R
+  Y <- Observed$Y
+  # Enkf---------------------------------------------------
+  ## process error
+  if (!is.null(Q)) {
+    Pf <- Pf + Q
+  }
+  
+  ## Kalman Gain
+  K <- Pf %*% t(H) %*% solve((R + H %*% Pf %*% t(H)))
+  # Analysis
+  mu.a <- mu.f + K %*% (Y - H %*% mu.f)
+  Pa   <- (diag(ncol(X)) - K %*% H) %*% Pf
+  return(list(mu.f = mu.f, Pf = Pf, mu.a = mu.a, Pa = Pa))
+}
+
+##' @title GEF
+##' @name  GEF
+##' @author Michael Dietze \email{dietze@@bu.edu}, Ann Raiho and Hamze Dokoohaki
+##' 
+##' @param settings  pecan standard settings list.  
+##' @param Forecast A list containing the forecasts variables including Pf (cov of forecast state variables), mu.f (vector of estimated mean of forecast state variables),
+##' Q (process variance) and X (a dataframe of forcats state variables for different ensemble)
+##' @param Observed A list containing the observed variables including R (cov of observed state variables) and Y (vector of estimated mean of observed state variables)
+##' @param ... Extra argument sent to the analysis function. This function requires an extra argument called `extraArg` that needs to be passed through ellipsis.
```

Comment:
As noted previously, extraArg should  be an explicit argument here, not passed via ellipsis.